core/src/main/scala/org/apache/spark/deploy/history/EventLogFileCompactor.scala

```diff
@@ -0,0 +1,250 @@
+ */
+class EventLogFileCompactor(
+    sparkConf: SparkConf,
+    hadoopConf: Configuration,
+    fs: FileSystem) extends Logging {
+
+  private val maxFilesToRetain: Int = sparkConf.get(EVENT_LOG_ROLLING_MAX_FILES_TO_RETAIN)
+
+  def compact(eventLogFiles: Seq[FileStatus]): Seq[FileStatus] = {
+    if (eventLogFiles.length <= maxFilesToRetain) {
+      return eventLogFiles
+    }
+
+    if (EventLogFileWriter.isCompacted(eventLogFiles.last.getPath)) {
+      return Seq(eventLogFiles.last)
+    }
+
+    val (filesToCompact, filesToRetain) = findFilesToCompact(eventLogFiles)
+    if (filesToCompact.isEmpty) {
+      filesToRetain
+    } else {
+      val bus = new ReplayListenerBus()
+
+      val builders = ServiceLoader.load(classOf[EventFilterBuilder],
+        Utils.getContextOrSparkClassLoader).asScala.toSeq
+      builders.foreach(bus.addListener)
+
+      filesToCompact.foreach { log =>
+        Utils.tryWithResource(EventLogFileReader.openEventLog(log.getPath, fs)) { in =>
+          bus.replay(in, log.getPath.getName)
+        }
+      }
+
+      val rewriter = new FilteredEventLogFileRewriter(sparkConf, hadoopConf, fs,
+        builders.map(_.createFilter()))
+      val compactedPath = rewriter.rewrite(filesToCompact)
+
+      cleanupCompactedFiles(filesToCompact)
+
+      fs.getFileStatus(new Path(compactedPath)) :: filesToRetain.toList
+    }
+  }
+
+  private def cleanupCompactedFiles(files: Seq[FileStatus]): Unit = {
+    files.foreach { file =>
+      var deleted = false
+      try {
+        deleted = fs.delete(file.getPath, true)
+      } catch {
+        case _: IOException =>
+      }
+      if (!deleted) {
+        logWarning(s"Failed to remove ${file.getPath} / skip removing.")
+      }
+    }
+  }
+
+  private def findFilesToCompact(
+      eventLogFiles: Seq[FileStatus]): (Seq[FileStatus], Seq[FileStatus]) = {
+    val files = RollingEventLogFilesFileReader.dropBeforeLastCompactFile(eventLogFiles)
+    if (files.length > maxFilesToRetain) {
+      (files.dropRight(maxFilesToRetain), files.takeRight(maxFilesToRetain))
+    } else {
+      (Seq.empty, files)
+    }
+  }
+}
+
+/**
+ * This class rewrites the event log files into one compact file: the compact file will only
+ * contain the events which pass the filters. Events will be filtered out only when all filters
+ * decide to filter out the event or don't mind about the event. Otherwise, the original line for
+ * the event is written to the compact file as it is.
+ */
+class FilteredEventLogFileRewriter(
+    sparkConf: SparkConf,
+    hadoopConf: Configuration,
+    fs: FileSystem,
+    filters: Seq[EventFilter]) extends Logging {
+
+  def rewrite(eventLogFiles: Seq[FileStatus]): String = {
+    require(eventLogFiles.nonEmpty)
+
+    val targetEventLogFilePath = eventLogFiles.last.getPath
+    val logWriter: CompactedEventLogFileWriter = new CompactedEventLogFileWriter(
+      targetEventLogFilePath, "dummy", None, targetEventLogFilePath.getParent.toUri,
```

Comment:
So that should be existing issue if the URI is the thing we don't feel safe, and I'm not totally sure how it would hurt.

> branch-2.4

https://github.com/apache/spark/blob/7bdc76f9c1f062aecdaf361527adbd9f50518bc9/core/src/main/scala/org/apache/spark/scheduler/EventLoggingListener.scala#L52-L58

I feel it worths to file a new issue and address from there, not only for event logging but also for every places.

Reply:
> I'm not totally sure how it would hurt

It can create directories which are not meant by the user. In the example the user expects directory/file `/.../chk %@#chk` but `/.../chk%20%25@%23chk` created. If the `toUri` is consistently used this is the "only" problem (for ex. https://github.com/apache/spark/pull/23733/files).

If in some places it's used but not on other places it can maybe SHS can't read things back which was written out (not checked just giving an example).

Reply:
Since this exists in previous versions it makes sense to handle it in a different jira to keep the focus.